pkg/codegen/hcl2/model/expression.go

```diff
@@ -1050,13 +1050,16 @@ func (x *FunctionCallExpression) Typecheck(typecheckOperands bool) hcl.Diagnosti
 	typecheckDiags := typecheckArgs(rng, x.Signature, x.Args...)
 	diagnostics = append(diagnostics, typecheckDiags...)
 
-	// Unless the function is already automatically using an
-	// Output-returning version, modify the signature to account
-	// for automatic lifting to Promise or Output.
-	_, isOutput := x.Signature.ReturnType.(*OutputType)
-	if !isOutput {
-		x.Signature.ReturnType = liftOperationType(x.Signature.ReturnType, x.Args...)
+	// Unless the function is already automatically using an Output-returning version, modify the signature to account
+	// for automatic lifting to Promise or Output. Note we skip this for pulumiResourceName and pulumiResourceType as
+	// they should return a plain string even with an output input.
+	if x.Name != "pulumi:pulumiResourceName" && x.Name != "pulumi:pulumiResourceType" {
+		_, isOutput := x.Signature.ReturnType.(*OutputType)
```

Comment:
This automatic lifting is _really_ suspect but removing it breaks a load of things. I'd like to revisit this and see if there's a cleaner way of doing it but this special case works for fixing the conformance tests up.